### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by bigspider**
- Added some documentation
- Refactored Watcher to remove StandardMappedComponent, and simplified the predicates
- Modified code of the BlockchainMachine to not assume that the state is computed if a block is in the BlockCache (false on startup)
- Slight simplification in BlockProcessor.

### CODE REVIEW

File: src/watcher/watcher.ts [hunk 1]
```diff
@@ -64,12 +64,26 @@ class AppointmentStateReducer implements StateReducer<WatcherAppointmentState, B
     }
 }
 
+/**
+ * Returns true if and only if the given predicate is true on `(state, block)` and false on `(oldState, oldBlock)`.
+ * TODO:198: could simplify this by removing the block parameter, and changing anchor states accordingly
+ */
+function predicateBecameTrue<TState, TBlock>(
```
Comment (yahgwai) [inline src/watcher/watcher.ts]: You could put this on the Component<T> class

Comment (bigspider) [reply src/watcher/watcher.ts]: Doesn't really work for MappedComponents, though, as the predicate is on single items, rather than on the whole state

Comment (bigspider) [reply src/watcher/watcher.ts]: Perhaps I can just export it in component.ts as a utility function?

Comment (yahgwai) [reply src/watcher/watcher.ts]: Do we have a mappedcomponent, I thought you'd removed that?

Comment (bigspider) [reply src/watcher/watcher.ts]: I removed the StandardMappedComponent, that also implemented handleNewStateEvents as a list of condition/action.
The MappedComponent just computes a reducer by mapping a uniform collection of object (string-indexed) to reducers for each object.

Comment (yahgwai) [reply src/watcher/watcher.ts]: Hmm, still not sure what you mean, I cant see a MappedComponent class on your branch.

Comment (bigspider) [reply src/watcher/watcher.ts]: Oh, right, we only have the MappedStateReducer. But still, Watcher extends Watcher extends Component<MappedState<WatcherAppointmentState>, Block>, while the preedicate is on WatcherAppointmentState only.

Comment (yahgwai) [reply src/watcher/watcher.ts]: OK, I guess the next best place to put it is on the anchor state itself

Comment (yahgwai) [reply src/watcher/watcher.ts]: In fact, we can create methods that match the predicates in the handleStateEvent function. And just call these directly on the state.

File: src/blockMonitor/blockchainMachine.ts [hunk 2]
```diff
@@ -54,23 +54,26 @@ export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopServi
...
 
                 states.set(block, state);
             }
 
             if (state && prevHead) {
-                const prevState = prevHead && states.get(prevHead)!;
-                // TODO:198: should we (deeply) compare old state and new state and only emit if different?
-                // Probably not, it might be expensive/inefficient depending on what is in TState
-                component.handleNewStateEvent(prevHead, prevState, head, state);
+                const prevState = prevHead && states.get(prevHead);
```
[the remark this answers is not in the document]
Comment (bigspider) [reply src/blockMonitor/blockchainMachine.ts]: Done in 1d9fc20.

File: src/blockMonitor/blockchainMachine.ts [hunk 3]
```diff
@@ -54,23 +54,26 @@ export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopServi
...
+                if (prevState != null) {
```
[the remark this answers is not in the document]
Comment (bigspider) [reply src/blockMonitor/blockchainMachine.ts]: Done in 1d9fc20.

File: src/blockMonitor/blockchainMachine.ts [hunk 4]
```diff
@@ -54,23 +54,26 @@ export class BlockchainMachine<TBlock extends IBlockStub> extends StartStopServi
         for (const { component, states } of this.componentsAndStates) {
             let state: object | null = null;
             for (const block of ancestorsToAdd) {
-                const parentBlock = this.blockProcessor.blockCache.getBlockStub(block.parentHash);
+                let prevAnchorState: object | undefined;
```
[the remark this answers is not in the document]
Comment (bigspider) [reply src/blockMonitor/blockchainMachine.ts]: Done in 1d9fc20.

**Pull request merged**